rfcs/0034-system-parameters.md

```diff
@@ -0,0 +1,135 @@
+
+- Changes to global parameters should be propagated to all nodes, in the cluster in a deterministic order ideally
+- When a node reboots or recovers, the latest snapshot of parameters should be passed from Meta Service.
+
+Therefore, the Notification Service seems to fit the scene well.
+
+### Session Parameters
+
+PostgreSQL allows users to set parameters for the current session as well.
+
+```sql
+SET parameter TO value / 'value' / DEFAULT;
+```
+
+References:
+
+- [PostgreSQL: Documentation: 15: 20.1. Setting Parameters](https://www.postgresql.org/docs/current/config-setting.html)
+
+Previously, we had already introduced [session parameters in RisingWave](https://github.com/risingwavelabs/risingwave/blob/53f7e0db772ac7e51773791bb8301624ed763ae8/src/common/src/session_config/mod.rs#L265). Some of them should be system parameters like `batch_enable_sort_agg` or `query_mode`. Postgres **enforces** that session parameters must also be system paramters, we can follow this rule as well.
+
+### Process Configs
+
+Let's rethink the current design of CLI options and config files. ([#5676](https://github.com/risingwavelabs/risingwave/issues/5676)) Given the existence of system parameters, some existing parameters may need to be moved to system parameters, like `data_directory` as mentioned before.
+
+We propose to consider CLI options as a way to override config files (i.e. in term of priority, CLI options -> config file -> defaults); as a result, the items in CLI options are a subset of the config file. The reasons for this include
+
+- All configs are still too many to be written as CLI options. (See [example.toml](https://github.com/risingwavelabs/risingwave/blob/main/src/config/example.toml)) 
+- While CLI configs are handy to set some frequently-used options like endpoint and memory.
+
+This pattern is widely adopted in many databases and even other applications. For example, TiDB considers configuration file as a superset of CLI options:
+
+> The TiDB configuration file supports more options than command-line parameters. You can download the default configuration file [config.toml.example](https://github.com/pingcap/tidb/blob/master/config/config.toml.example) and rename it to `config.toml`. This document describes only the options that are not involved in [command line options](https://docs.pingcap.com/tidb/stable/command-line-flags-for-tidb-configuration).
+>
+> Ref. [TiDB Configuration File | PingCAP Docs](https://docs.pingcap.com/tidb/stable/tidb-configuration-file)
+
+In RisingWave it will be like
+
+```toml
+# risingwave.toml
+[server]
+total_memory_bytes = 1073741824
+parallelism = 4
+[storage]
+shared_buffer_capacity_mb = 4096
+sstable_size_mb = 256
+block_size_kb = 1024
+```
+
+```bash
+# Override total_memory_bytes as 2GB
+./risingwave compute-node --total-memory-bytes=2147483648
+```
+
+### Mutability
+
+- Process configs must be **immutable**, because there is no user-interface to change it.
```

Comment:
I'm wondering if user has been running some workload for a while, and realize having a lower/higher `parallelism` will better utilize their resources.

In this scenario, `parallelism` is process level config, and is **immutable**.
What should the user do if they want to adjust it, do they have to start a new cluster?
Or is this not a realistic scenario that happens?

Reply:
> What should the user do if they want to adjust it, do they have to start a new cluster?

In that case, he/she can restart (or start new, they are equivalent actually) the CN process with new `parallelism` config. No need to restart the cluster.

-------

**Parallelism** is actually a very special case than other parameters. Let me tell the whole story in my mind:

Overall, we have

- Process-level **config** to control the parallelism of this CN node
- System **parameter** `parallelism` as the default parallelism for new streaming jobs. The default values is `0` or `'auto'` which implies the new streaming jobs are supposed to use all available parallel units in all compute nodes.
   - Sure, it can be session-level as well. 

When the system bootup:

1. Compute nodes register itself to Meta node when starting up, including its parallelism, which is controlled by the config `parallelism`
2. Meta service aggregates these numbers to get **total parallelism**, or we called the total number of parallel units. 
   - Note: "Parallel units" are basically one CPU core, but it's logical and is not strictly bound to a CPU core.
3. If the system parameter `parallelism` is set to `0` or `'auto'`, we will use that "total parallelism" in step 2 as the effective value to plan/schedule/run job actors.  Otherwise, the user-set value will be used.

If the cluster changed (e.g. new nodes joined or exiting nodes leaved), it will only affect the "total parallelism" in step 2.

Reply:
Basically, is it correct to say that `stream_parallelism` is only a meaningful config for dev/on-prem, to set a stream-job level default, but those using cloud should never have to care about it due to auto-redistribution?

Btw, I’m quite unclear, but do we currently have an API for scaling individual fragments? Is this API an SQL API or is it a HTTP/gRPC endpoint to meta?

Reply:
Also, is there a way for the user to hint the starting parallelism? For instance, even if cloud may auto distribute, it is unclear if it is better to start with small parallelism or high for a given stream job. User may know better.

So something like `create materialized view (stream_job_parallelism_hint=3) …` when the user thinks the stream job requires small throughput.

Reply:
> Btw, I’m quite unclear, but do we currently have an API for scaling individual fragments

Yes, See the design doc at https://singularity-data.quip.com/RubDA4a51oS0/Low-level-Scaling-Interface-on-Meta-Service

> So something like `create materialized view (stream_job_parallelism_hint=3)` … when the user thinks the stream job requires small throughput.

That's exactly the session variable `streaming_parallelism` does. See  https://github.com/risingwavelabs/risingwave/pull/7370. 
> but to me it is unclear what this variable does if we are going to auto scale, whereas a hint is at least clearer in that the parallelism may be modified by the system or by external operator.

I understand this seems to be somewhat inconsistent. In the current design, `streaming_parallelism` is just a hint for the subsequent `CREATE MVIEW` command. The syntax you proposed LGTM as well. After the MV is created, the only way to manipulate its parallelism is via the Scaling API (the [doc](https://singularity-data.quip.com/RubDA4a51oS0/Low-level-Scaling-Interface-on-Meta-Service) above). 

The reason is that we consider scaling as an advanced operational feature that **should be part of the cloud platform** instead of the database kernel, especially for the **policy** i.e. how and when to do scaling. Hence, we divide the feature "scaling" into 2 parts: 1) the cloud platform make decisions about how & when, and then 2) the kernel simply **executes** these directives. That's the design philosophy behind [that API](https://singularity-data.quip.com/RubDA4a51oS0/Low-level-Scaling-Interface-on-Meta-Service). However, the only exception is creating an MView, during which the meta must decide the parallelism without cloud platform engaged. In that case, I think the best we can do is to use `stream_parallelism` for each fragment.

Reply:
Can we say that local or process configs are immutable during runtime to avoid confusion? Because users can always kill and restart a process with different config values. 

> because there is no user-interface to change it.

That won't be true if the cloud portal supports updating with a reboot.